### PR TITLE
Glue component

### PR DESCRIPTION
**Pull request by TJJT777**
The glue component will prevent an entity from jumping and slow down movement by 10% for a specified duration of time.

### CODE REVIEW

File: src/main/java/org/terasology/alterationEffects/speed/GlueAlterationEffect.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package org.terasology.alterationEffects.speed;
+
+import org.terasology.alterationEffects.AlterationEffect;
+import org.terasology.alterationEffects.AlterationEffects;
+import org.terasology.context.Context;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.logic.delay.DelayManager;
+
+public class WalkSpeedAlterationEffect implements AlterationEffect {
```
Comment (Cervator) [inline src/main/java/org/terasology/alterationEffects/speed/GlueAlterationEffect.java]: Class name also needs to be `GlueAlterationEffect`

File: src/main/java/org/terasology/alterationEffects/speed/GlueAlterationEffect.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    private DelayManager delayManager;
+
+    public WalkSpeedAlterationEffect(Context context) {
```
Comment (Cervator) [inline src/main/java/org/terasology/alterationEffects/speed/GlueAlterationEffect.java]: Same for the constructor (change to `GlueAlterationEffect`)

File: src/main/java/org/terasology/alterationEffects/speed/GlueComponent.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+//done
```
Comment (Cervator) [inline src/main/java/org/terasology/alterationEffects/speed/GlueComponent.java]: Need the project license header here instead of `//done` (other files too)

File: src/main/java/org/terasology/alterationEffects/speed/SpeedAlterationSystem.java [hunk 4]
```diff
@@ -37,6 +37,10 @@ else if (event.getMovementMode() == MovementMode.WALKING && entityRef.hasCompone
         } else if (event.getMovementMode() == MovementMode.SWIMMING && entityRef.hasComponent(SwimSpeedComponent.class)) {
             event.multiply(entityRef.getComponent(SwimSpeedComponent.class).multiplier);
         }
+        
+        if (entityRef.hasComponent(GlueComponent.class)){
+            event.multiply(.9);
```
Comment (Cervator) [inline src/main/java/org/terasology/alterationEffects/speed/SpeedAlterationSystem.java]: The `.9` causes a compile error. `0.9f` works. The `f` is needed to create the number as a "float" data type.

File: src/main/java/org/terasology/alterationEffects/speed/GlueAlterationEffect.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        this.delayManager = context.get(DelayManager.class);
+    }
+
+    @Override
+    public void applyEffect(EntityRef instigator, EntityRef entity, float magnitude, long duration) {
+        boolean add = false;
+        GlueComponent glue = entity.getComponent(GlueComponent.class);
+        if (glue == null) {
+            add = true;
+            glue = new WalkSpeedComponent();
```
Comment (Cervator) [inline src/main/java/org/terasology/alterationEffects/speed/GlueAlterationEffect.java]: This doesn't work since `glue` is a `GlueComponent` and you're trying to assign a `WalkSpeedComponent` to it. Just need to update this to also use a `GlueComponent`

**Review by Cervator — CHANGES_REQUESTED** [review]
Hey @TJJT777 :-)

This looks pretty good, but also it appears like you're working without a full source workspace? There are compile errors in the code which should be pretty clear if you were working out of an IDE like IntelliJ. Additionally set up that way you can submit all the code in one commit rather than one per file/change.

Check out https://github.com/MovingBlocks/Terasology/wiki/Dev-Setup

If you get this set up in a source workspace the errors should show in the IDE, I've noted them here as well. In an IDE you can also actually test the change, like by swapping in the effect of an existing potion to use the glue effect instead :-)

**Comment by TJJT777** [comment]
Ok I fixed those changes. Ya those were pretty obvious I dont know how I didnt catch those.  I based my classes of the oldest one because it seemed as some of the newer ones didnt work but of course I couldnt tell because I am not coding inside an IDE.  So it should work now.

About running on intelliJ also,
when I run gradlew game it opens the game fine, so when I run that am I running it with the changes I make to the modules? Also there seems to be only 3 modules when I downloaded the file.  Where is the code being run from?  It cant be the three modules because there are only a few classes without much code.  So, when I make a change am I supposed to add for example the alteration module to the directory location of my the other modules then compile and run gradlew game?  Sorry the game dev help is just a little unclear on this(or maybe I am just not understanding it, not the gradlew game part but the part of how the code is being run) Thank You So Much!

**Comment by Cervator** [comment]
Thanks! I took it for a spin, although I realized I'd have to actually make a glue potion to test it. So now we have a glue potion as well over in the Potions module! :-)

For that I also had to add an entry for it in `EffectsAuthoritySystem` as well as do a little work over in Potions. While I was in there I fixed a few minor formatting issues, some of it pre-existing.

Changes added on top of this PR: https://github.com/Terasology/AlterationEffects/commit/b31966914c5e18f19d6ad3cc70343945c99c1308

Changes added on to Potions: https://github.com/Terasology/Potions/commit/e26cf589cfccfbbde3105499407c5109c49174ee

As for gradlew, IntelliJ, etc, there is a bunch of explanation in the game wiki, starting at: https://github.com/MovingBlocks/Terasology/wiki/Dev-Setup

In short yep you only get a few embedded modules with the main repo. Then you use commands like `gradlew fetchModuleAlterationEffects` to add that module to your workspace, and `gradlew idea` to update IntelliJ to know about the new module. At that point you can then both develop on the added module as well as more easily push things to GitHub :-)

`gradlew game` runs the game as well, with whatever modules are present locally, but IntelliJ gives you *a lot* more utility. Way easier to develop that way!

Please come on IRC any time to ask such questions, live chat is much easier for that part, but give the wiki a quick read first and go get yourself another task! Hope to see more from you soon :D

**Comment by TJJT777** [comment]
Wow feels great to have contributed. I will definitely by working on more tasks for terasology.  :) cheers
